tests/multigrid-global-coarsening/multigrid_a_03.cc

```diff
@@ -0,0 +1,153 @@
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+/**
+ * Test global-coarsening multigrid so that it works also on refinement levels.
+ */
+
+#include "multigrid_util.h"
+
+template <int dim, typename Number = double>
+void
+test(const unsigned int n_refinements, const unsigned int fe_degree_fine)
+{
+  using VectorType = LinearAlgebra::distributed::Vector<Number>;
+
+  parallel::distributed::Triangulation<dim> tria(
+    MPI_COMM_WORLD,
+    Triangulation<dim>::MeshSmoothing::none,
+    parallel::distributed::Triangulation<dim>::construct_multigrid_hierarchy);
+  GridGenerator::subdivided_hyper_cube(tria, 2);
+  tria.refine_global(n_refinements);
+
+  const unsigned int min_level = 0;
+  const unsigned int max_level = n_refinements;
+
+  MGLevelObject<AffineConstraints<Number>> constraints(min_level, max_level);
+  MGLevelObject<Operator<dim, Number>>     operators(min_level, max_level);
+
+  std::unique_ptr<FiniteElement<dim>> fe;
+  std::unique_ptr<Quadrature<dim>>    quad;
+  std::unique_ptr<Mapping<dim>>       mapping;
+
+  fe      = std::make_unique<FE_Q<dim>>(fe_degree_fine);
+  quad    = std::make_unique<QGauss<dim>>(fe_degree_fine + 1);
+  mapping = std::make_unique<MappingFE<dim>>(FE_Q<dim>(1));
+
+  // set up dofhandler
+  DoFHandler<dim> dof_handler(tria);
+  dof_handler.distribute_dofs(*fe);
+  dof_handler.distribute_mg_dofs();
+
+  const std::set<types::boundary_id> dirichlet_boundary = {0};
+  MGConstrainedDoFs                  mg_constrained_dofs;
+  mg_constrained_dofs.initialize(dof_handler);
+  mg_constrained_dofs.make_zero_boundary_constraints(dof_handler,
+                                                     dirichlet_boundary);
+
+  // set up levels
+  for (auto l = min_level; l <= max_level; ++l)
+    {
+      auto &constraint = constraints[l];
+      auto &op         = operators[l];
+
+      // set up constraints
+      IndexSet relevant_dofs;
+      DoFTools::extract_locally_relevant_level_dofs(dof_handler,
+                                                    l,
+                                                    relevant_dofs);
+      constraint.reinit(relevant_dofs);
+      constraint.add_lines(mg_constrained_dofs.get_boundary_indices(l));
+      constraint.close();
+
+      constraint.print(std::cout);
+
+      // set up operator
+      op.reinit(*mapping, dof_handler, *quad, constraint, l);
+    }
+
+  // set up transfer operator
+  MGTransferGlobalCoarsening<dim, VectorType> transfer(mg_constrained_dofs);
+
+  transfer.build(dof_handler, [&](const auto l, auto &vec) {
+    operators[l].initialize_dof_vector(vec);
+  });
+
+  GMGParameters mg_data; // TODO
```

Comment:
?

Reply:
The test is a copy of `multigrid_a_03` with the only difference that `MGConstrainedDoFs` is directly given to the transfer operators.